plugin/trino-hive/src/main/java/io/trino/plugin/hive/HivePartitionManager.java

```diff
@@ -111,26 +112,36 @@ public HivePartitionResult getPartitions(SemiTransactionalHiveMetastore metastor
                 .collect(toList());
 
         Iterable<HivePartition> partitionsIterable;
+        Optional<List<String>> partitionNames = hiveTableHandle.getPartitionNames();
         Predicate<Map<ColumnHandle, NullableValue>> predicate = constraint.predicate().orElse(value -> true);
         if (hiveTableHandle.getPartitions().isPresent()) {
             partitionsIterable = hiveTableHandle.getPartitions().get().stream()
                     .filter(partition -> partitionMatches(partitionColumns, effectivePredicate, predicate, partition))
                     .collect(toImmutableList());
         }
         else {
-            List<String> partitionNames = getFilteredPartitionNames(metastore, identity, tableName, partitionColumns, compactEffectivePredicate);
-            partitionsIterable = () -> partitionNames.stream()
+            List<String> partitionNamesList = hiveTableHandle.getPartitionNames()
+                    .orElse(getFilteredPartitionNames(metastore, identity, tableName, partitionColumns, compactEffectivePredicate));
+            partitionsIterable = () -> partitionNamesList.stream()
                     // Apply extra filters which could not be done by getFilteredPartitionNames
                     .map(partitionName -> parseValuesAndFilterPartition(tableName, partitionName, partitionColumns, partitionTypes, effectivePredicate, predicate))
                     .filter(Optional::isPresent)
                     .map(Optional::get)
                     .iterator();
+            partitionNames = Optional.of(partitionNamesList);
         }
 
-        // All partition key domains will be fully evaluated, so we don't need to include those
-        TupleDomain<ColumnHandle> remainingTupleDomain = effectivePredicate.filter((column, domain) -> !partitionColumns.contains(column));
-        TupleDomain<ColumnHandle> enforcedTupleDomain = effectivePredicate.filter((column, domain) -> partitionColumns.contains(column));
-        return new HivePartitionResult(partitionColumns, partitionsIterable, compactEffectivePredicate, remainingTupleDomain, enforcedTupleDomain, hiveBucketHandle, bucketFilter);
+        TupleDomain<ColumnHandle> remainingTupleDomain = effectivePredicate;
+        TupleDomain<ColumnHandle> enforcedTupleDomain = TupleDomain.all();
+
+        // Compute enforced and remaining TupleDomain if the partitions are loaded.
+        if (partitionNames.orElseGet(ImmutableList::of).size() <= maxPartitions || constraint.predicate().isPresent()) {
```

Comment:
`partitionNames` will be empty if `hiveTableHandle.getPartitions().isPresent()` is present, right? 
(according to javadoc in `HiveTableHandle)